libraries/standard/mqtt/src/mqtt.c

```diff
@@ -130,6 +234,447 @@ static int32_t sendPacket( MQTTContext_t * pContext,
 
 /*-----------------------------------------------------------*/
 
+static uint32_t calculateElapsedTime( uint32_t later, uint32_t start )
+{
+    return later - start;
```

Comment:
I was hoping this question would come up.

Short answer: This is to handle overflows, and I felt that it would be easy to forget to cast `later - start` into a `uint32_t` every time we want to calculate an elapsed time, so I made a function to return a `uint32_t`.

Long answer: According to [6.3.1.8](http://www.open-std.org/JTC1/sc22/wg14/www/docs/n1256.pdf) of the C standard, both values should get promoted to signed integers in the subtraction. I ran the following test on my machine:
```c
    uint8_t two = 2u, eight = 8u;

    printf("two - eight: %u\n", two - eight);
    if( (two - eight) < 0 )
        printf("two - eight < 0\n");
    else
        printf("two - eight >= 0\n");

    printf("(uint8_t) (two - eight): %u\n", (uint8_t) (two-eight));
    if((uint8_t) (two-eight) < 0)
        printf("(uint8_t) (two - eight) < 0\n");
    else
        printf("(uint8_t) (two - eight) >= 0\n");
```
and this was the result:
```
two - eight: 4294967290
two - eight < 0
(uint8_t) (two - eight): 250
(uint8_t) (two - eight) >= 0
```

So, in the case where the  `later` value has overflowed, and `later < start`, `later - start` will give us the correct answer (since 6.2.5.9 says that arithmetic with unsigned operands will never overflow, and be reduced modulo UINT_MAX), but only if it's converted back into a `uint32_t`. We could do this typecast every time we calculate the elapsed time, but I felt it might be easy to forget to do this, and so it would be better to make it a function.

Reply:
We need good unit tests.